qiskit/pulse/commands/_functional_pulse.py

```diff
@@ -0,0 +1,115 @@
+from qiskit.pulse.commands._sample_pulse import SamplePulse
+
+
+class FunctionalPulse:
+    """Decorator of pulse envelope function."""
+
+    def __init__(self, pulse):
+        """Create new pulse envelope function.
+
+        Args:
+            pulse (callable): a function describing pulse envelope.
+            pulse function must contain argument "duration".
+
+        Raises:
+            QiskitError: when incorrect envelope function is specified
+        """
+
+        if callable(pulse):
+            sig = signature(pulse)
+            if 'duration' in sig.parameters:
+                self.pulse = pulse
+            else:
+                raise QiskitError('Pulse function requires "duration" argument.')
+        else:
+            raise QiskitError('Pulse function is not callable.')
+
+    def __call__(self, duration, **kwargs):
+        """Create new functional pulse.
+        """
+        return FunctionalPulseCommand(self.pulse, duration=duration, **kwargs)
+
+
+class FunctionalPulseCommand(SamplePulse):
+    """Functional pulse."""
+
+    def __init__(self, pulse, duration, **kwargs):
+        """ Generate new pulse instance.
+
+        Args:
+            pulse (callable): a function describing pulse envelope
+            duration (int): duration of pulse
+        """
+
+        super(FunctionalPulseCommand, self).__init__(duration)
+
+        self.pulse = pulse
+        self._params = kwargs
+
+    @property
+    def params(self):
+        """Get parameters for describing pulse envelope
+
+        Returns:
+            dict: pulse parameters
+        """
+        return self._params
+
+    @params.setter
+    def params(self, params_new):
+        """Set parameters for describing pulse envelope
+
+        Args:
+            params_new (dict): dictionary of parameters
+        Raises:
+            QiskitError: when pulse parameter is not in the correct format.
+        """
+        if isinstance(params_new, dict):
+            for key, val in self._params.items():
+                self._params[key] = params_new.get(key, val)
+        else:
+            raise QiskitError('Pulse parameter should be dictionary.')
+
+    @property
+    def sample(self):
+        """Output pulse envelope as a list of complex values
+
+        Returns:
+            list: complex pulse envelope at each sampling point
+        Raises:
+            QiskitError: when invalid pulse data is generated
+        """
+        samples = self.pulse(self.duration, **self._params)
+
+        if not isinstance(samples, (list, np.ndarray)):
+            raise QiskitError('Output from pulse function is not array.')
+
+        if len(samples) != self.duration:
+            raise QiskitError('Number of Data point is not consistent with duration.')
+
+        if any(abs(samples) > 1):
+            warnings.warn("Absolute value of pulse amplitude exceeds 1.")
+            _samples = np.where(abs(samples) > 1, samples/abs(samples), samples)
```

Comment:
I think we should just raise there error and leave it up to the user to fix the underlying pulse function.